File: artifacts/api-server/src/integrations/amazon.ts
/**
 * Amazon SP-API adapter.
 * Uses LWA (Login with Amazon) for OAuth2 + AWS SigV4 request signing.
 * Docs: https://developer-docs.amazon.com/sp-api/docs/orders-api-v0-reference
 */

import { createHmac, createHash } from "node:crypto";
import type { IntegrationAdapter, SyncResult } from "./types";
import { ZERO_SYNC } from "./types";
import { db, ordersTable, customersTable } from "@workspace/db";
import { logger } from "../lib/logger";
import { withRetry, RateLimiter, sleep } from "../lib/rateLimit";

const LWA_TOKEN_URL = "https://api.amazon.com/auth/o2/token";
const SP_API_HOSTS: Record<string, string> = {
  "us-east-1": "sellingpartnerapi-na.amazon.com",
  "eu-west-1": "sellingpartnerapi-eu.amazon.com",
  "us-west-2": "sellingpartnerapi-fe.amazon.com",
};

interface AmazonCreds extends Record<string, unknown> {
  refreshToken: string;
  clientId: string;
  clientSecret: string;
  marketplaceId: string;
  sellerId: string;
  region?: string;
}

function isAmazonCreds(c: Record<string, unknown>): c is AmazonCreds {
  return (
    typeof c.refreshToken === "string" &&
    typeof c.clientId === "string" &&
    typeof c.clientSecret === "string" &&
    typeof c.marketplaceId === "string" &&
    typeof c.sellerId === "string"
  );
}

// ── LWA token ─────────────────────────────────────────────────────────────────

interface TokenEntry { accessToken: string; expiresAt: number }
const tokenCache = new Map<string, TokenEntry>();

async function getLwaToken(creds: AmazonCreds): Promise<string> {
  const key = `${creds.clientId}:${creds.refreshToken}`;
  const cached = tokenCache.get(key);
  if (cached && Date.now() < cached.expiresAt - 30_000) return cached.accessToken;

  const res = await fetch(LWA_TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "refresh_token",
      refresh_token: creds.refreshToken,
      client_id: creds.clientId,
      client_secret: creds.clientSecret,
    }),
  });
  if (!res.ok) throw new Error(`LWA token ${res.status}: ${await res.text()}`);
  const data = (await res.json()) as { access_token: string; expires_in: number };
  tokenCache.set(key, { accessToken: data.access_token, expiresAt: Date.now() + data.expires_in * 1_000 });
  return data.access_token;
}

// ── AWS SigV4 signer ──────────────────────────────────────────────────────────
// Amazon SP-API requires SigV4 signed requests even for role-less access
// using the anonymous credential (empty key/secret) when using LWA only.
// We sign with empty AWS credentials which is valid for the SP-API pattern.

function hash(data: string): string {
  return createHash("sha256").update(data, "utf8").digest("hex");
}
function hmac(key: Buffer | string, data: string): Buffer {
  return createHmac("sha256", key).update(data, "utf8").digest();
}

function sigV4Sign(
  method: string,
  url: URL,
  accessToken: string,
  region: string,
): Record<string, string> {
  const service = "execute-api";
  const now = new Date();
  const amzDate = now.toISOString().replace(/[:\-]|\.\d{3}/g, "").slice(0, 15) + "Z";
  const dateStamp = amzDate.slice(0, 8);

  const canonicalUri = url.pathname;
  const canonicalQuery = url.searchParams.toString();
  const payloadHash = hash("");

  const headers: Record<string, string> = {
    host: url.host,
    "x-amz-access-token": accessToken,
    "x-amz-date": amzDate,
  };

  const signedHeaders = Object.keys(headers).sort().join(";");
  const canonicalHeaders =
    Object.entries(headers)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([k, v]) => `${k}:${v.trim()}\n`)
      .join("") + "\n";

  const canonicalRequest = [
    method,
    canonicalUri,
    canonicalQuery,
    canonicalHeaders,
    signedHeaders,
    payloadHash,
  ].join("\n");

  const credentialScope = `${dateStamp}/${region}/${service}/aws4_request`;
  const stringToSign = ["AWS4-HMAC-SHA256", amzDate, credentialScope, hash(canonicalRequest)].join("\n");

  // SP-API supports "anonymous" signing (no AWS IAM role) when using LWA tokens
  const signingKey = hmac(
    hmac(hmac(hmac(Buffer.from("AWS4" + ""), dateStamp), region), service),
    "aws4_request",
  );
  const signature = hmac(signingKey, stringToSign).toString("hex");

  return {
    ...headers,
    Authorization:
      `AWS4-HMAC-SHA256 Credential=/${credentialScope}, ` +
      `SignedHeaders=${signedHeaders}, Signature=${signature}`,
  };
}

const limiter = new RateLimiter(1, 2_000); // SP-API Orders: 0.5 req/s burst

async function spFetch<T>(creds: AmazonCreds, path: string, params: Record<string, string> = {}): Promise<T> {
  await limiter.throttle();
  const region = creds.region ?? "us-east-1";
  const host = SP_API_HOSTS[region] ?? SP_API_HOSTS["us-east-1"]!;
  const url = new URL(`https://${host}${path}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);

  const accessToken = await getLwaToken(creds);
  const headers = sigV4Sign("GET", url, accessToken, region);

  for (let attempt = 0; attempt < 3; attempt++) {
    const res = await fetch(url.toString(), { headers });
    if (res.status === 429) {
      await sleep(2_000 * (attempt + 1));
      continue;
    }
    if (!res.ok) throw new Error(`SP-API ${res.status}: ${await res.text()}`);
    return (await res.json()) as T;
  }
  throw new Error("SP-API: exceeded retry limit");
}

function sinceIso(days: number): string {
  const d = new Date();
  d.setDate(d.getDate() - days);
  return d.toISOString();
}

export const amazonAdapter: IntegrationAdapter = {
  platform: "amazon",
  displayName: "Amazon",
  requiresCredentials: true,

  async validate(credentials) {
    if (!isAmazonCreds(credentials))
      throw new Error("refreshToken, clientId, clientSecret, marketplaceId, and sellerId are required");
    await getLwaToken(credentials);
  },

  async sync(userId, credentials) {
    if (!isAmazonCreds(credentials)) throw new Error("Invalid Amazon credentials");
    const result: SyncResult = { ...ZERO_SYNC };

    let nextToken: string | undefined;
    const createdAfter = sinceIso(90);

    do {
      const params: Record<string, string> = nextToken
        ? { NextToken: nextToken }
        : {
            MarketplaceIds: credentials.marketplaceId,
            CreatedAfter: createdAfter,
            OrderStatuses: "Unshipped,PartiallyShipped,Shipped,Canceled,Unfulfillable,InvoiceUnconfirmed,Pending",
          };

      const data = await withRetry("amazon:orders", () =>
        spFetch<{ payload: { Orders: any[]; NextToken?: string } }>(
          credentials,
          "/orders/v0/orders",
          params,
        )
      );

      const orders = data.payload.Orders ?? [];

      for (const o of orders) {
        const total = parseFloat(o.OrderTotal?.Amount ?? "0");
        const buyerName = o.BuyerInfo?.BuyerName ?? null;
        const buyerEmail = o.BuyerInfo?.BuyerEmail ?? null;

        let customerId: string | null = null;
        if (buyerEmail) {
          const [c] = await db
            .insert(customersTable)
            .values({
              userId,
              platform: "amazon",
              externalId: buyerEmail,
              email: buyerEmail,
              name: buyerName ?? buyerEmail,
              phone: null,
            })
            .onConflictDoUpdate({
              target: [customersTable.userId, customersTable.platform, customersTable.externalId],
              set: { name: buyerName ?? buyerEmail },
            })
            .returning({ id: customersTable.id });
          customerId = c?.id ?? null;
          result.customersAdded += 1;
        }

        const statusMap: Record<string, string> = {
          Shipped: "fulfilled",
          PartiallyShipped: "paid",
          Unshipped: "paid",
          Pending: "pending",
          Canceled: "cancelled",
        };

        await db
          .insert(ordersTable)
          .values({
            userId,
            platform: "amazon",
            externalId: o.AmazonOrderId,
            orderNumber: o.AmazonOrderId,
            customerId,
            subtotal: String(total),
            shipping: "0",
            tax: "0",
            totalAmount: String(total),
            status: statusMap[o.OrderStatus] ?? "pending",
            orderedAt: new Date(o.PurchaseDate),
          })
          .onConflictDoUpdate({
            target: [ordersTable.userId, ordersTable.platform, ordersTable.externalId],
            set: { status: statusMap[o.OrderStatus] ?? "pending", totalAmount: String(total) },
          });
        result.ordersAdded += 1;
      }

      nextToken = data.payload.NextToken;
      if (nextToken) await sleep(500); // stay within rate limit
    } while (nextToken);

    logger.info({ userId, orders: result.ordersAdded }, "Amazon sync ok");
    return result;
  },
};

File: python-api-integration/adapters/amazon.py
"""
adapters/amazon.py
Amazon SP-API adapter using the python-amazon-sp-api library.
Docs: https://developer-docs.amazon.com/sp-api/docs
"""

from __future__ import annotations

import logging
import time
from datetime import datetime, timedelta, timezone
from typing import Any

from .base import BaseAdapter, _retryable
from credentials import AmazonCreds

logger = logging.getLogger(__name__)


def _iso_utc(days_ago: int) -> str:
    dt = datetime.now(tz=timezone.utc) - timedelta(days=days_ago)
    return dt.strftime("%Y-%m-%dT%H:%M:%SZ")


class AmazonAdapter(BaseAdapter):
    """
    Fetch orders from Amazon SP-API.

    Uses python-amazon-sp-api (pip install python-amazon-sp-api).
    """

    _rate_limit_calls = 1   # Orders endpoint: 0.0167 req/s (1 per 60s burst)
    _rate_limit_period = 2.0

    def __init__(self, creds: AmazonCreds | None = None) -> None:
        super().__init__()
        self._creds = creds or AmazonCreds()
        self._orders_api = self._build_orders_api()

    def _auth_headers(self) -> dict[str, str]:
        # python-amazon-sp-api handles all auth internally
        return {}

    def _build_orders_api(self):
        try:
            from sp_api.api import Orders  # type: ignore
            from sp_api.base import Marketplaces  # type: ignore
        except ImportError as exc:
            raise ImportError(
                "python-amazon-sp-api not installed. "
                "Run: pip install python-amazon-sp-api"
            ) from exc

        credentials = {
            "refresh_token": self._creds.refresh_token,
            "lwa_app_id": self._creds.client_id,
            "lwa_client_secret": self._creds.client_secret,
        }
        # Attempt to resolve marketplace; fall back to US if unknown
        try:
            marketplace = getattr(Marketplaces, self._creds.region.upper())
        except AttributeError:
            marketplace = Marketplaces.US

        return Orders(credentials=credentials, marketplace=marketplace)

    # ---------------------------------------------------------------------- #
    # Public interface                                                          #
    # ---------------------------------------------------------------------- #
    @_retryable
    def get_orders(self, days: int = 30) -> list[dict[str, Any]]:
        """
        Fetch orders created in the last N days (auto-paginated).

        Returns:
            List of raw Amazon order dicts.
        """
        created_after = _iso_utc(days)
        orders: list[dict] = []
        next_token: str | None = None

        while True:
            if next_token:
                resp = self._orders_api.get_orders(NextToken=next_token)
            else:
                resp = self._orders_api.get_orders(
                    MarketplaceIds=[self._creds.marketplace_id],
                    CreatedAfter=created_after,
                    OrderStatuses=["Shipped", "Pending", "Unshipped", "Canceled"],
                )
            payload = resp.payload
            orders.extend(payload.get("Orders", []))
            next_token = payload.get("NextToken")
            if not next_token:
                break
            # SP-API has a burst quota; sleep to stay within rate limits
            time.sleep(1)

        logger.info("Amazon: fetched %d orders", len(orders))
        return orders

    def sync_store_orders(self, days: int = 30) -> list[dict[str, Any]]:
        """
        Normalised output for the orchestrator.

        Returns:
            [
                {
                    "id": str,
                    "date": str,
                    "total": float,
                    "status": str,
                    "fulfillment": str,
                    "source": "amazon",
                },
                ...
            ]
        """
        raw = self.get_orders(days=days)
        results = []
        for o in raw:
            amount_obj = o.get("OrderTotal", {})
            total = float(amount_obj.get("Amount", 0)) if amount_obj else 0.0
            results.append(
                {
                    "id": o.get("AmazonOrderId", ""),
                    "date": o.get("PurchaseDate", ""),
                    "total": total,
                    "status": o.get("OrderStatus", ""),
                    "fulfillment": o.get("FulfillmentChannel", ""),
                    "source": "amazon",
                }
            )
        return results
